samples/extensions/sparse_image/sparse_image.cpp

```diff
@@ -0,0 +1,1644 @@
+
+			vkCmdCopyBufferToImage(command_buffer, multi_page_buffer->get_handle(), virtual_texture.texture_image, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1U, &region);
+
+			virtual_texture.page_table[page_index].valid = true;
+		}
+		else
+		{
+			VkImageBlit blit_cmd{};
+			blit_cmd.srcOffsets[0]                 = {(block_offset.x) * 2, (block_offset.y) * 2, 0};
+			blit_cmd.srcOffsets[1]                 = {(block_offset.x + static_cast<int>(block_extent.width)) * 2, (block_offset.y + static_cast<int>(block_extent.height)) * 2, 1};
+			blit_cmd.srcSubresource.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;
+			blit_cmd.srcSubresource.mipLevel       = mip_level - 1U;
+			blit_cmd.srcSubresource.baseArrayLayer = 0U;
+			blit_cmd.srcSubresource.layerCount     = 1U;
+			blit_cmd.dstOffsets[0]                 = {block_offset.x, block_offset.y, 0};
+			blit_cmd.dstOffsets[1]                 = {block_offset.x + static_cast<int>(block_extent.width), block_offset.y + static_cast<int>(block_extent.height), 1};
+			blit_cmd.dstSubresource.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;
+			blit_cmd.dstSubresource.mipLevel       = mip_level;
+			blit_cmd.dstSubresource.baseArrayLayer = 0U;
+			blit_cmd.dstSubresource.layerCount     = 1U;
+
+			vkCmdBlitImage(command_buffer,
+			               virtual_texture.texture_image, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL,
+			               virtual_texture.texture_image, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL,
+			               1U, &blit_cmd,
+			               VK_FILTER_LINEAR);
+
+			virtual_texture.page_table[page_index].valid = true;
+		}
+	}
+	virtual_texture.update_set.clear();
+
+	VkImageSubresourceRange subresource_range{};
+
+	subresource_range.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;
+	subresource_range.baseArrayLayer = 0U;
+	subresource_range.layerCount     = 1U;
+	subresource_range.levelCount     = 1U;
+
+	if (current_mip_level != 0xFF)
+	{
+		subresource_range.baseMipLevel = current_mip_level;
+		vkb::image_layout_transition(command_buffer, virtual_texture.texture_image, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL, subresource_range);
+		if (current_mip_level != 0)
+		{
+			subresource_range.baseMipLevel = current_mip_level - 1U;
+			vkb::image_layout_transition(command_buffer, virtual_texture.texture_image, VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL, VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL, subresource_range);
+		}
+	}
+
+	device->flush_command_buffer(command_buffer, queue, true);
+
+	for (auto &page : virtual_texture.page_table)
+	{
+		page.gen_mip_required = false;
+	}
+}
+
+void SparseImage::free_unused_memory()
+{
+	for (size_t page_index = 0U; page_index < virtual_texture.page_table.size(); page_index++)
+	{
+		auto &page = virtual_texture.page_table[page_index];
+		if (page.render_required_set.empty() && page.valid)
+		{
+			page.valid  = false;
+			auto result = page.page_memory_info.memory_sector->available_offsets.insert(page.page_memory_info.offset);
+			if (!result.second)
```

Comment:
Maybe just `assert(result.second);`?

Reply:
Actually removed that statement (was used for corner-case testing and didn't occur at all).